### PR TITLE
Optimization/method bind optim and better bench

### PR DESCRIPTION
**Pull request by piiertho**
This improves icall performances caching `MethodBind` pointer.
This improves measures of benchmarks and add sufficient warmup iterations.
This improves benchmark results prints readbility.

### CODE REVIEW

File: harness/benchmarks/scripts/Main.gd [hunk 1]
```diff
@@ -1,59 +1,108 @@
...
+			__do_run(warmup, benchmark, stats, true)
+	for iteration in range(iterations):
+		__do_run(iteration, benchmark, stats, false)
+	printt("%s:" % benchmark.lang.name, str(stats))
+	report.add(benchmark, stats.get_results())
+
+
+func __do_run(iteration: int, benchmark: Benchmark, stats: Stats, is_warmup: bool):
+	var start: float = OS.get_ticks_usec()
+	benchmark.exec()
```
Comment (raniejade) [inline harness/benchmarks/scripts/Main.gd:59]: why not use iteration here as a loop?

File: harness/benchmarks/src/main/kotlin/godot/benchmark/Simple.kt [hunk 2]
```diff
@@ -54,20 +54,25 @@ class Simple : Object() {
...
+    fun benchmarkIcall(): Int {
         val node = Node()
         node.getInstanceId()
-        return node
+        node.free()
+        return 0
     }
 
-    fun benchmarkICallWithLoop(): Node {
+    fun benchmarkIcallWithLoop(): Int {
```
Comment (raniejade) [inline harness/benchmarks/src/main/kotlin/godot/benchmark/Simple.kt:64]: out of curiosity why the change to returning an int?

File: harness/benchmarks/scripts/Main.gd (hunk 1 above)
Comment (piiertho) [reply harness/benchmarks/scripts/Main.gd:59]: Because gdscript loops cost a lot as we can see in other bench. Si to avoid to mesure the cost of loop and so be more precise, we manually wrote calls.

File: harness/benchmarks/src/main/kotlin/godot/benchmark/Simple.kt (hunk 2 above)
Comment (piiertho) [reply harness/benchmarks/src/main/kotlin/godot/benchmark/Simple.kt:64]: Can be unit. Fix here is that we free node. Not freeing it was corrupting benchmark as GC was checking a list that was becoming bigger and bigger. That was causing cost augmentation in function of iterations number.

**Review by CedNaru — APPROVED** [review]

File: harness/benchmarks/gd/Simple.gd [hunk 3]
```diff
@@ -1,46 +1,51 @@
 func benchmark_simple_add():
-    var a = 1
-    var b = 2
-    return a + b
+	var a = 1
```
Comment (chippmann) [inline harness/benchmarks/gd/Simple.gd:2]: We need to setup our godot editors and IDE's to either use spaces or tabs for gdscript. Else we permanently have changes in those files.
Nothing to do with this particular PR but definitely something we have to look at sooner than later IMO

File: harness/benchmarks/gd/Simple.gd [hunk 4]
```diff
@@ -1,46 +1,51 @@
 func benchmark_simple_add():
```
Comment (chippmann) [inline harness/benchmarks/gd/Simple.gd:1]: Shouldn't we now use strong typing in gdscript as well? IIRC the perf benefits of typed gdscript have been backported to 3.2 and will be available with the next patch release. IMO it would be more fair for gdscript to benefit from that in the benchmark.
Best would be to have two gdscript benchmarks: one typed and one not.

File: harness/benchmarks/scripts/Benchmarks.gd [hunk 5]
```diff
@@ -1,23 +1,53 @@
...
+		scripts.append([cls.new(), method_list, lang])
+		
+		for method_dict in method_dict_list:
+			var method_name = method_dict["name"]
+			
+			#check the method name stats with "benchmark_" so we don't test methods like "init"
+			if method_name.begins_with("benchmark_"):
+				method_list.append(method_name)
+	
+				#If not already in the global method last, we add it.
```
Comment (chippmann) [inline harness/benchmarks/scripts/Benchmarks.gd:33]: typo: list not last

File: harness/benchmarks/scripts/Benchmarks.gd [hunk 6]
```diff
@@ -1,23 +1,53 @@
...
+				if !methods.has(method_name):
+					methods.append(method_name)
+				
+	
+	methods.sort()
+	print("Tests to run:")
+	##Loop over the methods, then the scripts so they are all next toeach other for a same test.
```
Comment (chippmann) [inline harness/benchmarks/scripts/Benchmarks.gd:40]: typo: toeach -> to each

File: harness/benchmarks/scripts/Main.gd (hunk 1 above)
Comment (chippmann) [reply harness/benchmarks/scripts/Main.gd:59]: Understandable. But maybe add a comment explaining this?

File: harness/benchmarks/scripts/Stats.gd [hunk 7]
```diff
@@ -1,66 +1,69 @@
...
\ No newline at end of file
+	var results = get_results()
+	var _min = results.min
+	var _max = results.max
+	var avg = results.avg
+	var median = results.median
+	var pc05 = results.p05
+	var pc95 = results.p95
+	var args = [avg, _min, pc05, median, pc95, _max]
+	return "AVG=%d op/s, MIN=%d op/s, 05PC=%d op/s, MEDIAN=%d op/s, 95PC=%d op/s, MAX=%d op/s" % args
```
Comment (chippmann) [inline harness/benchmarks/scripts/Stats.gd:69]: could be use tabs between the results rather than spaces? so the numbers are evenly spaced for all benchmark methods? This would greatly improve readability IMO

File: kt/godot-runtime/src/main/kotlin/godot/core/GarbageCollector.kt [hunk 8]
```diff
@@ -19,6 +19,8 @@ object GarbageCollector {
 
     private var forceJvmGarbageCollector = false
 
+    var shouldDisplayLeakInstancesOneClose = true
```
Comment (chippmann) [inline kt/godot-runtime/src/main/kotlin/godot/core/GarbageCollector.kt:22]: Typo: shouldDisplayLeakInstancesOneClose -> shouldDisplayLeakInstancesOnClose

File: src/gd_kotlin.cpp [hunk 9]
```diff
@@ -154,6 +167,9 @@ void GDKotlin::init() {
             is_gc_activated = false;
             //TODO: Link to documentation
             WARN_PRINT("GC thread was disable. --jvm-disable-gc should only be used for debugging purpose")
+        } else if (cmd_arg == "--jvm-disable-closing-leaks-warning") {
```
Comment (chippmann) [inline src/gd_kotlin.cpp:170]: Could we add a little doc to the readme about all the cmdline arguments we can pass in?

Comment (chippmann) [reply src/gd_kotlin.cpp:170]: Because there are quite a few now

**Review by chippmann — CHANGES_REQUESTED** [review]

**Comment by chippmann** [comment]
@CedNaru & @piiertho can you please double check that the csharp benchmarks now really are similar to the other languages?

File: harness/benchmarks/gd/Simple.gd (hunk 3 above)
Comment (piiertho) [reply harness/benchmarks/gd/Simple.gd:2]: I agree with this, on my side I got conflicts between intellij and godot editor. Would be simpler to make a PR to solve this issue.

File: src/gd_kotlin.cpp (hunk 9 above)
Comment (piiertho) [reply src/gd_kotlin.cpp:170]: If you're ok I'll do it in a seperate PR, in which I will create a conf object on cpp side so that we can move command args parsing from GdKotlin singleton. This would improve maintainability. Maybe write an issue about this ?

**Comment by piiertho** [comment]
> @CedNaru & @piiertho can you please double check that the csharp benchmarks now really are similar to the other languages?

Ok on my side :)

File: harness/benchmarks/gd/Simple.gd (hunk 3 above)
Comment (chippmann) [reply harness/benchmarks/gd/Simple.gd:2]: Ok for me

File: src/gd_kotlin.cpp (hunk 9 above)
Comment (chippmann) [reply src/gd_kotlin.cpp:170]: Ok I'll write one and assign it to you

File: harness/benchmarks/gd/Simple.gd (hunk 4 above)
Comment (piiertho) [reply harness/benchmarks/gd/Simple.gd:1]: @CedNaru ?

**Comment by piiertho** [comment]
> @CedNaru & @piiertho can you please double check that the csharp benchmarks now really are similar to the other languages?

@CedNaru ?

File: harness/benchmarks/scripts/Main.gd (hunk 1 above)
Comment (piiertho) [reply harness/benchmarks/scripts/Main.gd:59]: I agree with this, @CedNaru can you fix benchmark, it was more your modifications IIRC.

Comment (CedNaru) [reply harness/benchmarks/scripts/Main.gd:59]: Done

File: harness/benchmarks/scripts/Benchmarks.gd (hunk 5 above)
Comment (CedNaru) [reply harness/benchmarks/scripts/Benchmarks.gd:33]: done

File: harness/benchmarks/scripts/Stats.gd (hunk 7 above)
Comment (CedNaru) [reply harness/benchmarks/scripts/Stats.gd:69]: done

File: harness/benchmarks/scripts/Benchmarks.gd (hunk 6 above)
Comment (CedNaru) [reply harness/benchmarks/scripts/Benchmarks.gd:40]: done

File: harness/benchmarks/scripts/Stats.gd (hunk 7 above)
Comment (CedNaru) [reply harness/benchmarks/scripts/Stats.gd:69]: image
Using tabs gives us ugly results. As the tab push the next string to a multiple of 4 spaces so it doesn't align properly.

File: harness/benchmarks/gd/Simple.gd (hunk 4 above)
Comment (CedNaru) [reply harness/benchmarks/gd/Simple.gd:1]: Done. But there is no difference between Typed and Dynamic yet. Those changes are only for Godot 4.0

image

**Comment by CedNaru** [comment]
image

Total is going to be inferred as an integer, so a integer division is going to be used at the end. It's unfair because Integer division is more expensive than a float divison.

**Comment by CedNaru** [comment]
I changed the icall return type to void in gdscript, it should probably be done the same for the other languages.

File: harness/benchmarks/scripts/Stats.gd (hunk 7 above)
Comment (chippmann) [reply harness/benchmarks/scripts/Stats.gd]: Ok let's leave it then

File: harness/benchmarks/typed_gd/Simple.gd [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+		var v: Vector3 = Vector3(i, i, i)
+		v = b.xform(v)
+		s = s + v
+
+	return s
+
+func benchmark_vectors2_only() -> Vector2:
+	var s: Vector2 = Vector2()
+	for i in range(1000):
+		var v = Vector2(i, i)
```
Comment (chippmann) [inline harness/benchmarks/typed_gd/Simple.gd:29]: @CedNaru Not typed. Either type it explicitly or use `:=`

**Review by chippmann — CHANGES_REQUESTED** [review]

**Comment by chippmann** [comment]
> I changed the icall return type to void in gdscript, it should probably be done the same for the other languages.

Changed kotlin and csharp to do the same. @piiertho and @CedNaru , please look over that as well real quick :-)

File: harness/benchmarks/typed_gd/Simple.gd (hunk 10 above)
Comment (CedNaru) [reply harness/benchmarks/typed_gd/Simple.gd]: done

File: harness/benchmarks/gd/Simple.gd (hunk 4 above)
Comment (chippmann) [reply harness/benchmarks/gd/Simple.gd:1]: Cool thx. IIRC they are in the process of being backported to 3.x (but i can't find the PR anymore). But maybe I'm just wrong. Anyways good to have it.

**Review by chippmann — APPROVED** [review]

**Comment by piiertho** [comment]
> > I changed the icall return type to void in gdscript, it should probably be done the same for the other languages.
> 
> Changed kotlin and csharp to do the same. @piiertho and @CedNaru , please look over that as well real quick :-)

seems ok to me

**Review by CedNaru — APPROVED** [review]
